mlflow/store/db_migrations/versions/cfd24bdc0731_update_run_status_constraint_with_killed.py

```diff
@@ -0,0 +1,52 @@
+"""Update run status constraint with killed
+
+Revision ID: cfd24bdc0731
+Revises: 89d4b8295536
+Create Date: 2019-10-11 15:55:10.853449
+
+"""
+from alembic import op, context
+from mlflow.store.tracking.dbmodels.models import SqlRun
+from mlflow.entities import RunStatus
+
+# revision identifiers, used by Alembic.
+revision = 'cfd24bdc0731'
+down_revision = '2b4d017a5e9b'
+branch_labels = None
+depends_on = None
+
+
+new_run_status = [
+    RunStatus.to_string(RunStatus.SCHEDULED),
+    RunStatus.to_string(RunStatus.FAILED),
+    RunStatus.to_string(RunStatus.FINISHED),
+    RunStatus.to_string(RunStatus.RUNNING),
+    RunStatus.to_string(RunStatus.KILLED)
+]
+
+previous_run_status = [
+    RunStatus.to_string(RunStatus.SCHEDULED),
+    RunStatus.to_string(RunStatus.FAILED),
+    RunStatus.to_string(RunStatus.FINISHED),
+    RunStatus.to_string(RunStatus.RUNNING)
+]
+
+
+def _is_backend_sqlite():
+    return context.get_bind().engine.url.get_backend_name() == 'sqlite'
+
+
+def upgrade():
+    with op.batch_alter_table("runs") as batch_op:
+        if not _is_backend_sqlite():
+            batch_op.drop_constraint(constraint_name='status', type_='check')
```

Comment:
This constraint may not exist against MySQL databases running version 8.0.15 or earlier; check constraints were introduced in MySQL 8.0.16. Attempting to run this migration against 8.0.15 yields the following error output:
```
> mlflow db upgrade mysql://root:password@localhost:33060/checkconstraint

2019/12/06 17:45:12 INFO mlflow.store.db.utils: Updating database tables at mysql://root:password@localhost:33060/checkconstraint
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 2b4d017a5e9b -> cfd24bdc0731, Update run status constraint with killed
Traceback (most recent call last):
  File "/Users/czumar/anaconda2/bin/mlflow", line 11, in <module>
    load_entry_point('mlflow', 'console_scripts', 'mlflow')()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/czumar/mlflow/mlflow/db.py", line 29, in upgrade
    mlflow.store.db.utils._upgrade_db(url)
  File "/Users/czumar/mlflow/mlflow/store/db/utils.py", line 123, in _upgrade_db
    command.upgrade(config, 'heads')
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/command.py", line 298, in upgrade
    script.run_env()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/script/base.py", line 489, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/util/pyfiles.py", line 98, in load_python_file
    module = load_module_py(module_id, path)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/util/compat.py", line 239, in load_module_py
    mod = imp.load_source(module_id, path, fp)
  File "/Users/czumar/mlflow/mlflow/store/db_migrations/env.py", line 78, in <module>
    run_migrations_online()
  File "/Users/czumar/mlflow/mlflow/store/db_migrations/env.py", line 72, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/runtime/environment.py", line 846, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/runtime/migration.py", line 518, in run_migrations
    step.migration_fn(**kw)
  File "/Users/czumar/mlflow/mlflow/store/db_migrations/versions/cfd24bdc0731_update_run_status_constraint_with_killed.py", line 44, in upgrade
    condition=SqlRun.status.in_(new_run_status))
  File "/Users/czumar/anaconda2/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/operations/base.py", line 325, in batch_alter_table
    impl.flush()
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/operations/batch.py", line 79, in flush
    fn(*arg, **kw)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/ddl/mysql.py", line 112, in drop_constraint
    super(MySQLImpl, self).drop_constraint(const)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/ddl/impl.py", line 241, in drop_constraint
    self._exec(schema.DropConstraint(const))
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/alembic/ddl/impl.py", line 134, in _exec
    return conn.execute(construct, *multiparams, **params)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/sql/ddl.py", line 72, in _execute_on_connection
    return connection._execute_ddl(self, multiparams, params)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1050, in _execute_ddl
    compiled,
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1248, in _execute_context
    e, statement, parameters, cursor, context
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1466, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 383, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
    cursor, statement, parameters, context
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/MySQLdb/cursors.py", line 206, in execute
    res = self._query(query)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/MySQLdb/cursors.py", line 312, in _query
    db.query(q)
  File "/Users/czumar/anaconda2/lib/python2.7/site-packages/MySQLdb/connections.py", line 224, in query
    _mysql.connection.query(self, query)
sqlalchemy.exc.ProgrammingError: (MySQLdb._exceptions.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'CHECK status' at line 1")
[SQL: ALTER TABLE runs DROP CHECK status]
(Background on this error at: http://sqlalche.me/e/f405)
```

Is it possible to check for the presence of this constraint and drop it only if it previously existed?

Reply:
Done